File: scripts/toolbox/python/sirene.py
#!/usr/bin/env python3
"""
sirene.py — accéder proprement au répertoire Sirene.

Trois stratégies, selon le besoin (voir la Boîte à outils du site) :

  stock   analyses de masse, population complète. Le script interroge les
          MÉTADONNÉES data.gouv du jeu Sirene pour retrouver la ressource
          la plus récente (CSV zippé ou Parquet) : aucun lien mensuel n'est
          figé dans le code.
  api     requêtes ciblées (une unité, un critère). Nécessite une clé du
          portail api.insee.fr dans la variable d'environnement
          SIRENE_API_TOKEN (en-tête X-INSEE-Api-Key-Integration).
  geo     géolocalisation des établissements pour études territoriales.

Exemples :
    python scripts/toolbox/python/sirene.py stock --dry-run
    python scripts/toolbox/python/sirene.py stock --fichier StockUniteLegale --format parquet --dry-run
    python scripts/toolbox/python/sirene.py api --siren 552100554 --dry-run
    python scripts/toolbox/python/sirene.py geo --dry-run

Sans --dry-run, `stock` et `geo` téléchargent dans data/raw/ (ignoré par
git). `api` exécute la requête (quelques Ko).
"""

from __future__ import annotations

import argparse
import os
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent))
import helpers  # noqa: E402

DATAGOUV_API = "https://www.data.gouv.fr/api/1/datasets"
DATASET_SIRENE = "base-sirene-des-entreprises-et-de-leurs-etablissements-siren-siret"
DATASET_GEO = ("geolocalisation-des-etablissements-du-repertoire-sirene-"
               "pour-les-etudes-statistiques")
SIRENE_API = "https://api.insee.fr/api-sirene/3.11"


def latest_resource(dataset: str, name_contains: str,
                    fmt: str | None = None) -> dict:
    """Retrouve la ressource la plus récente d'un jeu data.gouv.

    On lit les métadonnées du dataset plutôt que de figer une URL de
    fichier mensuel : le lien reste valable millésime après millésime.
    """
    meta = helpers.http_json(f"{DATAGOUV_API}/{dataset}/")
    candidates = [
        r for r in meta.get("resources", [])
        if name_contains.lower() in (r.get("title") or "").lower()
        and (fmt is None or (r.get("format") or "").lower() == fmt.lower())
    ]
    if not candidates:
        raise SystemExit(
            f"Aucune ressource « {name_contains} » (format={fmt}) dans le "
            f"jeu {dataset} : vérifier les métadonnées sur data.gouv.fr.")
    candidates.sort(key=lambda r: r.get("last_modified") or "", reverse=True)
    return candidates[0]


def cmd_stock(args) -> None:
    fmt = None if args.format == "auto" else args.format
    # « StockEtablissement - » et non « StockEtablissement » : le jeu contient
    # aussi StockEtablissementLiensSuccession et StockEtablissementHistorique,
    # qu'un simple test de sous-chaîne attraperait par erreur.
    resource = latest_resource(DATASET_SIRENE, f"{args.fichier} -", fmt)
    print(f"ressource la plus récente : {resource['title']}")
    print(f"  modifiée le : {(resource.get('last_modified') or '?')[:10]}")
    dest = helpers.RAW_DIR / Path(resource["url"]).name
    helpers.download(resource["url"], dest, dry_run=args.dry_run)
    if args.dry_run:
        print("\nEnsuite : lire en colonnes projetées avec DuckDB, jamais "
              "en chargement complet (voir la Boîte à outils, recette 03).")


def cmd_api(args) -> None:
    token = os.getenv("SIRENE_API_TOKEN")
    url = f"{SIRENE_API}/siret?q=siren:{args.siren}&nombre={args.nombre}"
    headers = {"X-INSEE-Api-Key-Integration": token or "(SIRENE_API_TOKEN)"}
    if args.dry_run:
        helpers.describe_request(
            f"établissements du SIREN {args.siren} via l'API Sirene 3.11",
            url, headers=headers)
        if not token:
            print("  note     : SIRENE_API_TOKEN absent de l'environnement "
                  "(clé du portail api.insee.fr)")
        return
    if not token:
        raise SystemExit("Définir SIRENE_API_TOKEN (clé du portail "
                         "api.insee.fr) — jamais de clé dans le code.")
    payload = helpers.http_json(url, headers=headers)
    for etab in payload.get("etablissements", []):
        adresse = etab.get("adresseEtablissement", {})
        print(etab.get("siret"),
              etab.get("uniteLegale", {}).get("denominationUniteLegale", ""),
              adresse.get("libelleCommuneEtablissement", ""))


def cmd_geo(args) -> None:
    resource = latest_resource(DATASET_GEO, "geolocalisation")
    print(f"ressource la plus récente : {resource['title']}")
    dest = helpers.RAW_DIR / Path(resource["url"]).name
    helpers.download(resource["url"], dest, dry_run=args.dry_run)


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    sub = parser.add_subparsers(dest="commande", required=True)

    p_stock = sub.add_parser("stock", help="fichiers stock mensuels")
    p_stock.add_argument("--fichier", default="StockEtablissement",
                         choices=["StockEtablissement", "StockUniteLegale"])
    p_stock.add_argument("--format", default="auto",
                         choices=["auto", "parquet", "csv"],
                         help="parquet disponible depuis 2026 sur data.gouv")
    p_stock.add_argument("--dry-run", action="store_true")
    p_stock.set_defaults(func=cmd_stock)

    p_api = sub.add_parser("api", help="requête ciblée API Sirene 3.11")
    p_api.add_argument("--siren", required=True)
    p_api.add_argument("--nombre", type=int, default=20,
                       help="pagination : nombre de résultats par appel")
    p_api.add_argument("--dry-run", action="store_true")
    p_api.set_defaults(func=cmd_api)

    p_geo = sub.add_parser("geo", help="géolocalisation des établissements")
    p_geo.add_argument("--dry-run", action="store_true")
    p_geo.set_defaults(func=cmd_geo)

    args = parser.parse_args()
    args.func(args)


if __name__ == "__main__":
    main()

File: scripts/toolbox/R/sirene.R
# sirene.R — accéder proprement au répertoire Sirene depuis R.
#
# Stratégie « stock » : on interroge les MÉTADONNÉES data.gouv du jeu
# Sirene pour retrouver la ressource la plus récente (CSV zippé ou
# Parquet) — aucun lien mensuel figé dans le code.
#
# Dépendances : httr2 (install.packages("httr2")).
# Usage :
#   Rscript scripts/toolbox/R/sirene.R stock --dry-run
#   Rscript scripts/toolbox/R/sirene.R stock StockUniteLegale parquet
#   Rscript scripts/toolbox/R/sirene.R api 552100554 --dry-run
#
# La clé de l'API Sirene 3.11 (requêtes ciblées) vient de la variable
# d'environnement SIRENE_API_TOKEN — jamais du code.

source(file.path(dirname(sub("--file=", "",
  grep("--file=", commandArgs(), value = TRUE)[1])), "helpers.R"))

DATASET <- paste0("https://www.data.gouv.fr/api/1/datasets/",
                  "base-sirene-des-entreprises-et-de-leurs-etablissements-",
                  "siren-siret/")
SIRENE_API <- "https://api.insee.fr/api-sirene/3.11"

derniere_ressource <- function(motif, format = NULL) {
  meta <- http_json(DATASET)
  ressources <- Filter(function(r) {
    # « StockEtablissement - » et non « StockEtablissement » : le jeu
    # contient aussi les fichiers LiensSuccession et Historique.
    ok <- grepl(paste0(motif, " -"), r$title %||% "", ignore.case = TRUE)
    if (!is.null(format)) ok <- ok && identical(tolower(r$format %||% ""),
                                                tolower(format))
    ok
  }, meta$resources)
  if (length(ressources) == 0) stop("Aucune ressource « ", motif, " » : ",
                                    "vérifier les métadonnées data.gouv.")
  dates <- vapply(ressources, function(r) r$last_modified %||% "", "")
  ressources[[order(dates, decreasing = TRUE)[1]]]
}

`%||%` <- function(x, y) if (is.null(x)) y else x

cmd_stock <- function(fichier, format, dry_run) {
  r <- derniere_ressource(fichier, format)
  cat("ressource la plus récente :", r$title, "\n")
  telecharger(r$url, file.path(RAW_DIR, basename(r$url)), dry_run = dry_run)
  if (dry_run) cat("\nEnsuite : lecture en colonnes projetées avec duckdb",
                   "ou arrow, jamais read.csv() complet.\n")
}

cmd_api <- function(siren, dry_run) {
  url <- sprintf("%s/siret?q=siren:%s&nombre=20", SIRENE_API, siren)
  if (dry_run) {
    decrire_requete(paste("établissements du SIREN", siren), url,
                    list(`en-tête` = "X-INSEE-Api-Key-Integration = ***"))
    return(invisible(NULL))
  }
  token <- Sys.getenv("SIRENE_API_TOKEN")
  if (!nzchar(token)) stop("Définir SIRENE_API_TOKEN (portail api.insee.fr).")
  reponse <- http_json(url, list(`X-INSEE-Api-Key-Integration` = token))
  for (etab in reponse$etablissements) {
    cat(etab$siret, "|",
        etab$adresseEtablissement$libelleCommuneEtablissement %||% "", "\n")
  }
}

args <- commandArgs(trailingOnly = TRUE)
dry_run <- "--dry-run" %in% args
args <- setdiff(args, "--dry-run")

if (length(args) == 0) stop("Commande attendue : stock | api")
if (args[[1]] == "stock") {
  cmd_stock(fichier = if (length(args) >= 2) args[[2]] else "StockEtablissement",
            format = if (length(args) >= 3) args[[3]] else NULL,
            dry_run = dry_run)
} else if (args[[1]] == "api") {
  if (length(args) < 2) stop("Usage : api <siren> [--dry-run]")
  cmd_api(args[[2]], dry_run)
} else {
  stop("Commande inconnue : ", args[[1]])
}
